hal/src/thumbv7em/clock/v2.rs

```diff
@@ -221,25 +223,67 @@ pub fn retrieve_clocks(
     }
 }
 
-/// Marker supertrait unifying family of more specific source marker traits.
+/// Marks clock types that can drive downstream clocks
 ///
-/// These ones are essential during a construction (`fn ::{new, enable}`) and
+/// Implementers of this type can drive downstream clocks in the clock tree.
+/// Typically, implementors are [`Enabled`] clocks. The `Source` associated type
+/// maps to the corresponding `Id` type of the implementer.
 ///
-/// deconstruction (`fn ::{free, disable}`) of clocking components as they
-/// provide information to the constructed/deconstructed type what its source is
-/// and which variant of source (associated constant) is applicable while
-/// performing a HW register write.
-pub trait SourceMarker: crate::typelevel::Sealed {}
+/// For example, `Enabled<Gclk5<DfllId>>` would implement
+/// `Driver<Source = GclkId5>`.
+pub trait Driver: Sealed {
```

Comment:
After taking some time I think now understand fully how this change is supposed to work. You moved the duplicated layering of `Gclk/Dpll/PclkSource` and `Gclk/Dpll/PclkSourceMarker` by removing and squashing the former into the `Source` (renamed into `Driver`) and removing the associated type constraint. Matching against the correct source is done within `impl` blocks by using `*Id` types. Earlier the matching was redundant. Very nice job! I got too narrow minded to pick such a thing up.

Reply:
Precisely. I didn't recognize it either until I read that forum post and started thinking about redundant trait bounds in several different places.

Do we like `Driver`? Or should we go back to `Source`? If we go back to `Source`, I think it should be
```rust
pub trait Source: Sealed {
    type Id;
    fn freq(&self) -> Hertz;
}
```

I don't have particularly strong feelings. If I can try to restate my previous post briefly, I think I was aiming for:
> I **am** a driver. I drive other clocks. So I implement `Driver`

vs.

> I **receive** a clock source. So I look up the `Source` of my `Driver`.

I'm not sure if that distinction helps or not. I thought it did, but I think `Source` and `Id` would also work well.

Reply:
`Source` and `Id` feels better for me but that's aesthetics.

Reply:
I think I'll switch to `Source` and `Id`. Now that there are fewer traits using the word `Source`, it feels less overloaded to me. I also like that `Id` clearly indicates that the type should be one of the `Id` types.